### PR TITLE
[systemtest] Documentation generator for tests

### PR DESCRIPTION
**Pull request by im-konge**
### Type of change

- Enhancement / new feature

### Description

This PR adds a generator for the STs documentation, which is present in the new module - `systemtest-doc`.
The generator, based on the parameters, generates Markdown file for each test class and for each method adds record to the file. To achieve that, new annotations are added -> `@TestDoc`, `@TestDoc.Step`, `@TestDoc.Usecase`, `@TestDoc.Desc`.

Description of each annotation and how the generation works are mentioned in the `README.md` file inside the `systemtest-doc` module.

Additionally, this PR adds new profile to the `systemtest` module, which help us with copying all the needed dependencies to the `target` file of the module, that can be then added to the classpath of the Java exec command.

For this PR, I added an example to `ListenersST#testSendMessagesPlainAnonymous` and the generated file can be seen in `systemtest-doc/documentation/io/strimzi/systemtest/kafka/listeners/ListenersST.md`.

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [X] Update documentation

### CODE REVIEW

**Comment by scholzj** [comment]
Wouldn't a separate repo be better for this?

File: pom.xml [hunk 1]
```diff
@@ -268,6 +269,12 @@
                 <artifactId>kafka-oauth-keycloak-authorizer</artifactId>
                 <version>${strimzi-oauth.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.strimzi</groupId>
+                <artifactId>systemtest-doc</artifactId>
+                <version>${project.version}</version>
+                <scope>compile</scope>
+            </dependency>
```
Comment (scholzj) [inline pom.xml:277]: Is this really needed if it is part of this repo?

**Review by scholzj — COMMENTED** [review]
How do you expect this to work? Where do you expect the generted files to be? How do you check if they are up to date etc.?

**Comment by im-konge** [comment]
> Wouldn't a separate repo be better for this?

For the generator or the ST documentation?

I was thinking about it and the generator needs the built classes of the `systemtest` module + all the dependencies, so I decided to keep it in this repository.

> How do you expect this to work?

The build is not forced during the `mvn clean install`, by which it would do some kind of cyclic dependency.
When we edit the documentation of test, we need to manually generate it (I guess it is similarly done in case of the `api` module, but correct me if I'm wrong).

>  Where do you expect the generted files to be?

Currently it is generated inside the `systemtest-doc` module into the `documentation` directory.

> How do you check if they are up to date etc.?

Checking if there are changes in the test that is already documented would be non-trivial FMPOV. Because there are lot of things that can happen and that we would have to watch. Instead, we can add some check that would print some information that there are changes to the test and that the contributor should update the documentation and re-generate it. In the case of new tests, it should be relatively easy to make the build (or some PR check) fail, with an information that the test doesn't contain the documentation.

File: pom.xml (hunk 1 above)
Comment (im-konge) [reply pom.xml:277]: Will check it, thanks :)

**Review by jankalinic — APPROVED** [review]

**Comment by scholzj** [comment]
> For the generator or the ST documentation?

For the generator. The ST documentation has to live with the STs. I think there are several reasons for it:
* The lifecycle of the generator does not seem to be tied to Strimzi releases
* It seems like something usable outside this repo => for example in the Drain Cleaner repo, but possible even outside of Strimzi
* It will help to keep the build times in check as well
* It can be probably easily integrated for example as a Maven plugin or similar

> Checking if there are changes in the test that is already documented would be non-trivial FMPOV. Because there are lot of things that can happen and that we would have to watch. Instead, we can add some check that would print some information that there are changes to the test and that the contributor should update the documentation and re-generate it. In the case of new tests, it should be relatively easy to make the build (or some PR check) fail, with an information that the test doesn't contain the documentation.

If you plan to commit the docs into the repo, you should have some checks for it. Similar to how we check the files in docs and packaging for example.

**Comment by im-konge** [comment]
@scholzj yeah that makes sense. Should I create a proposal for adding a new repository to the organization?

**Comment by scholzj** [comment]
You can. But I would even consider having it outside of Strimzi. This seems to be generally useful - and assuming you checked and there is no similar tool that already exists, I think it might be interesting even outside of Strimzi.

**Comment by im-konge** [comment]
Okay, I will move it outside of this repository and keep the documentation and the build here. I'll just convert it to draft for now.

Thanks! :)

**Comment by scholzj** [comment]
Discussed on the community call on 4th May: Should be closed and will be reopened later.

**Pull request closed without merging**